### PR TITLE
add unzip

### PR DESCRIPTION
**Pull request by adavies42**
THE FOLLOWING DISCLAIMER APPLIES TO ALL SOFTWARE CODE AND OTHER MATERIALS CONTRIBUTED IN CONNECTION WITH THIS SOFTWARE:
THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.
ONLY THE SOFTWARE CODE AND OTHER MATERIALS CONTRIBUTED IN CONNECTION WITH THIS SOFTWARE, IF ANY, THAT ARE ATTACHED TO (OR OTHERWISE ACCOMPANY) THIS SUBMISSION (AND ORDINARY COURSE CONTRIBUTIONS OF FUTURES PATCHES THERETO) ARE TO BE CONSIDERED A CONTRIBUTION. NO OTHER SOFTWARE CODE OR MATERIALS ARE A CONTRIBUTION.

### CODE REVIEW

**Review by alvin-c-shih — APPROVED** [review]

File: q/unzip/unzip.q [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+
+
+// Utilities
+
+// Read bytes from either a file or a byte vector.
+// @param x hsym or bytes
+// @param y offset
+// @param z length
+// @return z bytes from x, starting at y
+.finos.unzip.priv.bytes:{$[-11h=t:type x;read1(x;y;z);4h=t;x y+til z;'`type]}
```
Comment (gyorokpeter) [inline q/unzip/unzip.q:11]: How does the `til` method of slicing an array compare to take+drop for large sizes? Or is this supposed to be used for small sizes only?

File: q/unzip/unzip.q [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+// Parse a range of data with a header.
+// parser is a function of three arguments:
+//  Its first argument will be (data;extra); extra is passed as :: if not
+//   included.
+//  Its second argument will be the starting index of the record to extract.
+//  Its third argument will be the raw headers of the record, split and
+//   labeled according to fields.
+//  It should return (record;next index).
+// parser will be called until it returns next index equal to length.
+// @param x (parser;fields;extra)
```
Comment (gyorokpeter) [inline q/unzip/unzip.q:52]: I think using multiple implicit parameters becomes confusing if the function is longer than a line or two.

File: q/util/util.q [hunk 3]
```diff
@@ -13,3 +13,75 @@
 // @return A byte vector containing the contents of the file.
 // @see read1
 .finos.util.read1f:{r:{y,read1 x}[h:hopen`$":fifo://",1_string x]over();hclose h;r}
+
+.finos.util.compose:('[;])/
+
+// create a list. e.g. list(`a;1) -> (`a;1)
+.finos.util.list:{$[104h=type x;1_-1_get x;x]}
```
Comment (gyorokpeter) [inline q/util/util.q:20]: This needs more examples for why it is useful (also considering we already have a function to create a list, it's called `enlist`). For most types it returns the parameter unchanged, which means if it's an atom the result is not a list, contrary to what the description suggests. The only special case is for projections, for which we drop the function that is being projected, so we only keep the bound arguments... but then we also drop the last of those arguments?

File: q/unzip/unzip.q (hunk 1 above)
Comment (adavies42) [reply q/unzip/unzip.q:11]: i haven't actually timed it, but take+drop involves an intermediate list which is longer than the index list created by `til`, so i wouldn't expect it to be significantly better, if any

File: q/unzip/unzip.q [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+
+// Split a subsection of data into fields.
+// Starts from offset and takes sum fields entries, splitting them according.
+//  to fields.
+// fields is a dictionary of field names and widths.
+// @param x fields
+// @param y offset
+// @param z data
+// @return the split subsection of the vector
+.finos.unzip.priv.split:{(key x)!(get sums prev x)cut z y+til sum x}
```
[the remark this answers is not in the document]
Comment (adavies42) [reply q/unzip/unzip.q:26]: whose convention? i always use `get` for both

File: q/unzip/unzip.q (hunk 2 above)
Comment (adavies42) [reply q/unzip/unzip.q:52]: this entire file is written using implicit pa

File: q/util/util.q (hunk 3 above)
Comment (adavies42) [reply q/util/util.q:20]: the point is to let you write in "trailing delimiter" style,

```
foo:(
    1;
    2;
    )
```

which has advantages for source control and general future-proofing

i can add an example to that effect

it's also used by `util.dict` and `util.table` to provide the same feature to them

File: q/util/util.q [hunk 5]
```diff
@@ -13,3 +13,75 @@
...
+.finos.util.table:{flip x!flip(count x)cut .finos.util.list y}
+
+// log stubs
+.finos.log.critical:{-1"CRITICAL: ",x;}
+.finos.log.error   :{-1"ERROR: "   ,x;}
+.finos.log.warning :{-1"WARNING: " ,x;}
+.finos.log.info    :{-1"INFO: "    ,x;}
+.finos.log.debug   :{-1"DEBUG: "   ,x;}
+
+.finos.util.rs  :{0b sv x xprev 0b vs y}     / right shift
```
[the remark this answers is not in the document]
Comment (adavies42) [reply q/util/util.q:35]: fair, i'll change that

File: q/unzip/unzip.q (hunk 4 above)
[the remark this answers is not in the document]
Comment (gyorokpeter) [reply q/unzip/unzip.q:26]: https://code.kx.com/q/ref/value/#get

[the remark this answers is not in the document]
Comment (adavies42) [reply q/unzip/unzip.q:26]: "but the two are interchangeable" :)

**Comment by gyorokpeter** [comment]
For the fun of it, I also added a zip schema to [qbinparse](https://github.com/gyorokpeter/qbinparse/blob/master/schema/zip.sch). It's not enough to actually unzip files (since schemas don't contain business logic) but it help break down the various header structures.

**Pull request merged**